File: app/src/main/java/com/moisesborges/tvaddict/episodes/EpisodesPresenter.java
package com.moisesborges.tvaddict.episodes;

import android.support.annotation.NonNull;

import com.moisesborges.tvaddict.data.ShowsRepository;
import com.moisesborges.tvaddict.models.Episode;
import com.moisesborges.tvaddict.models.Show;
import com.moisesborges.tvaddict.mvp.BasePresenter;
import com.moisesborges.tvaddict.mvp.RxJavaConfig;

import java.util.ArrayList;
import java.util.List;
import java.util.Timer;

import javax.inject.Inject;

import io.reactivex.Observable;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.BiFunction;
import timber.log.Timber;


/**
 * Created by Moisés on 21/04/2017.
 */

public class EpisodesPresenter extends BasePresenter<EpisodesView> {


    private final ShowsRepository mShowsRepository;

    @Inject
    public EpisodesPresenter(@NonNull RxJavaConfig rxJavaConfig,
                             @NonNull ShowsRepository showsRepository) {
        super(rxJavaConfig);
        mShowsRepository = showsRepository;
    }

    public void loadEpisodes(int seasonNumber,
                             @NonNull Show show) {
        checkView();

        Disposable disposable = mShowsRepository.getSavedShow(show.getId())
                .compose(applySchedulersToSingle())
                .onErrorReturn(ignored -> Show.NOT_FOUND)
                .doOnSuccess(showFromDb -> {
                    if (showFromDb != Show.NOT_FOUND) {
                        getView().setShow(showFromDb);
                    }
                })
                .map(showFromDb -> showFromDb != Show.NOT_FOUND ? showFromDb.getEpisodes() : show.getEpisodes())
                .map(episodes -> {
                    List<Episode> episodesFromSeason = new ArrayList<>();
                    for (Episode episode : episodes) {
                        if (episode.getSeason() == seasonNumber) {
                            episodesFromSeason.add(episode);
                        }
                    }
                    return episodesFromSeason;
                })
                .subscribe(episodes -> getView().displayEpisodes(episodes), Timber::e);

        addDisposable(disposable);
    }

    public void changeEpisodeSeenStatus(@NonNull Episode episode,
                                        @NonNull Show show) {
        checkView();

        episode.setWatched(!episode.wasWatched());
        Disposable disposable = mShowsRepository.updateShow(show)
                .compose(applySchedulersToSingle())
                .subscribe(ignored -> getView().refreshEpisode(episode), Timber::e);

        addDisposable(disposable);
    }
}

File: app/src/main/java/com/moisesborges/tvaddict/episodes/EpisodesPresenter.kt
package com.moisesborges.tvaddict.episodes

import com.moisesborges.tvaddict.data.ShowsRepository
import com.moisesborges.tvaddict.models.Episode
import com.moisesborges.tvaddict.models.Show
import com.moisesborges.tvaddict.mvp.BasePresenter
import com.moisesborges.tvaddict.mvp.RxJavaConfig

import javax.inject.Inject

import timber.log.Timber


/**
 * Created by Moisés on 21/04/2017.
 */

class EpisodesPresenter
@Inject constructor(rxJavaConfig: RxJavaConfig,
                    private val showsRepository: ShowsRepository) : BasePresenter<EpisodesView>(rxJavaConfig) {

    fun loadEpisodes(seasonNumber: Int, show: Show) {
        checkView()

        val disposable = showsRepository.getSavedShow(show.id!!)
                .compose(applySchedulersToSingle<Show>())
                .onErrorReturn { ignored -> Show.NOT_FOUND }
                .doOnSuccess { showFromDb ->
                    if (showFromDb !== Show.NOT_FOUND) {
                        view.setShow(showFromDb)
                    }
                }
                .map { showFromDb -> if (showFromDb !== Show.NOT_FOUND) showFromDb.episodes else show.episodes }
                .map<List<Episode>> { episodes -> episodes.filter { episode -> episode.season == seasonNumber } }
                .subscribe({ episodes -> view.displayEpisodes(episodes) }, { Timber.e(it) })

        addDisposable(disposable)
    }

    fun changeEpisodeSeenStatus(episode: Episode, show: Show) {
        checkView()

        if (!episode.wasAired()) {
            return
        }

        episode.setWatched(!episode.wasWatched())
        val disposable = showsRepository.updateShow(show)
                .compose(applySchedulersToSingle<Show>())
                .subscribe({ ignored -> view.refreshEpisode(episode) }, { Timber.e(it) })

        addDisposable(disposable)
    }

    fun openEpisodeDetails(episode: Episode) {
        checkView()

        view.showEpisodeDetails(episode)
    }

    fun setAllEpisodesStatusAsSeen(seasonNumber: Int, show: Show) {
        setAllEpisodesSeenStatus(show, seasonNumber, true)

    }

    private fun setAllEpisodesSeenStatus(show: Show, seasonNumber: Int, watched: Boolean) {
        checkView()

        show.episodes
                .filter { episode -> episode.season == seasonNumber }
                .filter { episode -> episode.wasAired() }
                .forEach { episode -> episode.setWatched(watched) }

        val disposable = showsRepository.updateShow(show)
                .compose(applySchedulersToSingle<Show>())
                .subscribe({ ignored -> view.refreshAllEpisodes() }, { Timber.e(it) })

        addDisposable(disposable)
    }

    fun setAllEpisodesStatusAsUnseen(seasonNumber: Int, show: Show) {
        setAllEpisodesSeenStatus(show, seasonNumber, false)
    }


}
